### ISSUE TITLE
Questions about adapting IRT for other datasets

### ISSUE THREAD

**Issue by chaitanyamalaviya** [nyu-mll/nlu-test-sets#4]
Hi, thanks for making your code available!

I had a couple of questions:
1) How can I extend your codebase to perform evaluation on a held-out set of responses? After looking through the code, it doesn't seem obvious to me how one could do this, and I was looking for any suggestions you might have.
2) Running your codebase on another set of datasets throws the following error. I already verified that the format/structure of the response files are valid. Do you have any suggestions for how to resolve this?

Thanks!!

```
Traceback (most recent call last):
  File "/scratch/cmalaviya/nlu-test-sets/irt_scripts/variational_irt.py", line 609, in <module>
    main(args)
  File "/scratch/cmalaviya/nlu-test-sets/irt_scripts/variational_irt.py", line 479, in main
    _ = train(
  File "/scratch/cmalaviya/nlu-test-sets/irt_scripts/variational_irt.py", line 321, in train
    elbo_loss = svi_kernel.step(weights, data_)
  File "/scratch-local/cmalaviya/nlu-test-sets/irt_scripts/weighted_ELBO.py", line 173, in step
    loss = self.loss_and_grads(self.model, self.guide, weights, *args, **kwargs)
  File "/scratch-local/cmalaviya/nlu-test-sets/irt_scripts/weighted_ELBO.py", line 115, in loss_and_grads
    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/pyro/infer/elbo.py", line 170, in _get_traces
    yield self._get_trace(model, guide, args, kwargs)
  File "/scratch-local/cmalaviya/nlu-test-sets/irt_scripts/weighted_ELBO.py", line 30, in _get_trace
    model_trace, guide_trace = get_importance_trace(
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/pyro/infer/enum.py", line 44, in get_importance_trace
    guide_trace = poutine.trace(guide, graph_type=graph_type).get_trace(*args, **kwargs)
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/pyro/poutine/trace_messenger.py", line 187, in get_trace
    self(*args, **kwargs)
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/pyro/poutine/trace_messenger.py", line 171, in __call__
    raise exc from None
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/pyro/poutine/trace_messenger.py", line 165, in __call__
    ret = self.fn(*args, **kwargs)
  File "/scratch/cmalaviya/nlu-test-sets/irt_scripts/variational_irt.py", line 286, in <lambda>
    guide = lambda obs: vi_posterior(obs, alpha_dist, theta_dist)
  File "/scratch/cmalaviya/nlu-test-sets/irt_scripts/variational_irt.py", line 154, in vi_posterior
    dist.Normal(
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/torch/distributions/normal.py", line 50, in __init__
    super(Normal, self).__init__(batch_shape, validate_args=validate_args)
  File "/scratch/cmalaviya/anaconda3/lib/python3.8/site-packages/torch/distributions/distribution.py", line 53, in __init__
    raise ValueError("The parameter {} has invalid values".format(param))
ValueError: The parameter loc has invalid values
Trace Shapes:
 Param Sites:
         b mu 124569
     b logstd 124569
Sample Sites:
```

**Comment by sleepinyourhat** [comment]
@claravania @phu-pmh, mind taking a look?

**Comment by phu-pmh** [comment]
The procedure for adding new test/val resposes is mentioned [here](https://github.com/nyu-mll/nlu-test-sets#analyzing-new-test-sets).
- To extend the experiments on a new test or held-out data, the responses for the new dataset should be placed under the `data` folder. The response file should be named `{dataset-name}__irt_all_coded.csv`  and it should have the same format as other response files under `data` folder.
- Add the `{dataset-name}` to the `TASKS` parameter in [`estimate_irt_params.sh`](https://github.com/nyu-mll/nlu-test-sets/blob/main/irt_scripts/estimate_irt_params.sh#L17) file.
- If needed, you can change IRT parameters such as the distribution type (normal or lognormal), the STD for the alpha, beta, and theta in `estimate_irt_params.sh` using `DISTS`, `ALPH_STDS`, and `PARAM_STDS` variables.
- Run the model using `bash irt_scripts/estimate_irt_params.sh` command.

The error might be because `item_param_std` is not set properly. If it doesn't solve the problem, could you provide us the command you used for running `variational_irt.py`?

**Comment by chaitanyamalaviya** [comment]
Thanks but I actually wanted to ask how one would evaluate the trained IRT model on pairs of (responder, response) that have not been seen during training of the IRT model. I basically wanted to estimate the fit of an IRT model on held-out responses. 

As for setting `item_param_std`, I'm just running `estimate_irt_params.sh` with the default values in the script, while just changing the task names like the `README.md` says. Would you recommend searching over other values of this hyperparameter?

**Comment by phu-pmh** [comment]
Thanks for clarifying. Our code unfortunately does not support the inference on the unseen datasets so you would need to add your own inference code.

I'm not sure where the error is coming from but seems like there is a problem with the values of the difficulty parameter (`b mu` and `b logstd`) according to the error trace. It might be good to search over other values.

**Comment by chaitanyamalaviya** [comment]
Thanks. The reliance on `pyro` made it a bit hard to find the best way to do this, since I'm not familiar with the package. For eg, it seems like the trace of the response matrix is computed using pyro, which I think might not be possible if some values are held-out and hence, missing. 

Got it, thanks! In the paper, you mention that you used `item_param_std=1.0`. What values did you search over before deciding on this value?

**Comment by phu-pmh** [comment]
Yes, we fixed the `item_param_std=1.0` in our experiments and only varied alpha -- it worked for us. Were you having the error while replicating our experiments or while adding a new dataset? Could you also check if `n_models, n_items` in `irt_model` and `vi_posterior` are the same?

**Comment by chaitanyamalaviya** [comment]
I was getting the error when trying to fit a model with a different set of datasets. I actually managed to resolve it by trying other values for `item_param_std` (in the range [0.1, 1000] with increments of 10x).
Thanks again!
